### DISCUSSION TITLE
help debugging code for a psychophysical experiment

### DISCUSSION THREAD

**Discussion by mingjia1025** [jspsych/jsPsych#1031] [Q&A] [answered]
Are there anyone familiar with the psychophysics plug-in [http://jspsychophysics.hes.kyushu-u.ac.jp/](url)? I'm having trouble running the main part of my experiment in javascript. Can someone help me debug my code (attached as a txt file)
[my code.txt](https://github.com/jspsych/jsPsych/files/5180856/my.code.txt)

**Comment by mingjia1025** [comment]
When I run my code through a html page, I get a grey bar in the middle of the webpage and the following error:

`Uncaught TypeError: Cannot convert undefined or null to object

    at Function.keys (<anonymous>)
    at evaluateTimelineVariables (jspsych.js:863)
    at doTrial (jspsych.js:831)
    at startExperiment (jspsych.js:784)
    at jspsych.js:2569
    at Object.jsPsych.pluginAPI.module.preloadVideo (jspsych.js:2449)
    at jspsych.js:2568
    at Object.jsPsych.pluginAPI.module.preloadAudioFiles (jspsych.js:2295)
    at jspsych.js:2567
    at Object.jsPsych.pluginAPI.module.preloadImages (jspsych.js:2400)`

**Comment by jodeleeuw** [comment]
@kurokida may be in the best position to help on this. Tagging him in case he hasn't seen it.

**Comment by kurokida** [comment, marked as answer]
Thank you for using my plugin.

Unfortunately, I am unable to determine the cause at this time.

The error code seems to inform that files could not be preloaded.

Can I see the complete source code?
Or can I have access to the program of the experiment?

**Reply by kurokida** [reply]
Sorry, now I noticed the attachment file.

I noticed a part of the code that needs to be fixed.

stimuli: rect_object, 

should be,

stimuli: [rect_object], 

Please see the section "4. Specify a trial object including the stimuli in the jsPsych’s timeline" at http://jspsychophysics.hes.kyushu-u.ac.jp/

**Reply by mingjia1025** [reply]
I made some changes to my code. Now it seems like a deep issue is about defining some parameters of rect_object as timelineVariables rather than fixed values as in your demonstration code.  When I ran my code, I was asked to specify the width, height, fill_color and line_color, all of which are coded as timelineVariables that are defined later at the trainingBlock function. 

Do you know how to solve this problem?


[my code.txt](https://github.com/jspsych/jsPsych/files/5189824/my.code.txt)

**Reply by mingjia1025** [reply]
@kurokida @jodeleeuw  Can you help with my error in setting up timelineVariables in the psychophysics functions? Thanks!

**Reply by jodeleeuw** [reply]
In most cases timeline variables only can be used as a replacement for the top-level parameter. Here the parameter `stimuli` is the top-level parameter, and all of the objects contained within it cannot be timeline variables in the way that you've set it up. What you could do is redefine your timeline variables to match the structure of the `stimuli` parameter, i.e., have a single variable that contains the array of stimuli objects.

**Reply by mingjia1025** [reply]
@jodeleeuw  @kurokida  I modified my code as you suggested and now I can iterate through my trials and blocks with different rect_objects . However, even though I can see a rectangle of correct side_length and color in the classification part of a trial, in the following feedback part, I can only see a rectangle of the same color but of shrunk size located in the lower right corner. **_(my timeline parameter has a list of 2 psychophysics plug-ins, both of which need to use 'rect_object' as a stimuli parameter so that the same rectangle is presented in each trial.)_** After the first block of my trials, I can see no rectangle in either part of a trial.  I'm not sure of the cause of that wired behavior. Can you help me fix it?

[my code.txt](https://github.com/jspsych/jsPsych/files/5212118/my.code.txt)

**Reply by kurokida** [reply]
I'm very sorry. I found a bug regarding the jspsych-psychophysics plugin.
The attachment file is the latest modified version.
Could you check that your program runs properly using the latest file?

[jspsych-psychophysics.zip](https://github.com/jspsych/jsPsych/files/5212339/jspsych-psychophysics.zip)

**Reply by mingjia1025** [reply]
Thanks! It's much better now. One minor issue: When the stimulus is large, I have to scroll down to see the full prompt. Is there a way to control the size of the gap between stimuli and prompt?

**Reply by kurokida** [reply]
The size of the canvas is the same as the size of the window when you don't specify the canvas_width and canvas_height property. As a result, the prompt is presented off-screen.

There are two ways to resolve the issue.

One is to specify the canvas_height property as with [the demonstration program](https://www.hes.kyushu-u.ac.jp/~kurokid/jspsychophysics/demos/draw_two_images_repeatedly.html).

The other is to use a text object instead of the prompt. The text object can be made by specifying the obj_type: ‘text’, and must be included in the stimuli property of the psychophysics plugin.
Please see [the demonstration program](https://www.hes.kyushu-u.ac.jp/~kurokid/jspsychophysics/demos/texts.html).

**Comment by kurokida** [comment]
Dear , @jodeleeuw ,

I have a question.
As I wrote above, the problem has been solved by changing my jspsych-psychophysics plugin code.
But, now I feel like this problem is better solved not on my code but on the jspsych code.

Let me explain.

In my plugin, a lot of properties regarding stimuli (e.g., position, angle, and color) are changed during the trial.
I noticed that these changed properties are inherited the next time they are called.
That is, these properties are not reset to their initial values experimenters defined.

I think this problem will occur even when other plugins were used.

I'm not sure if this is a bug in jspsych or not, but it would be very helpful if you could solve this problem on the jspsych code.
